apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputScores/PlaygroundOutputScores.tsx

```diff
@@ -0,0 +1,356 @@
+import React, { useMemo, useRef, useEffect } from "react";
+import { Loader2 } from "lucide-react";
+
+import useTraceById from "@/api/traces/useTraceById";
+import useRulesList from "@/api/automations/useRulesList";
+import useProjectByName from "@/api/projects/useProjectByName";
+import useAppStore from "@/store/AppStore";
+import FeedbackScoreTag from "@/components/shared/FeedbackScoreTag/FeedbackScoreTag";
+import {
+  HoverCard,
+  HoverCardContent,
+  HoverCardTrigger,
+} from "@/components/ui/hover-card";
+import { cn } from "@/lib/utils";
+import { TraceFeedbackScore } from "@/types/traces";
+import { PLAYGROUND_PROJECT_NAME } from "@/constants/shared";
+import { TAG_VARIANTS_COLOR_MAP } from "@/components/ui/tag";
+import { generateTagVariant } from "@/lib/traces";
+import { EvaluatorsRule, EVALUATORS_RULE_TYPE } from "@/types/automations";
+
+const MAX_VISIBLE_METRICS = 3;
+
+// Helper to get score names that a rule will produce
+const getScoreNamesFromRule = (rule: EvaluatorsRule): string[] => {
+  if (
+    rule.type === EVALUATORS_RULE_TYPE.llm_judge ||
+    rule.type === EVALUATORS_RULE_TYPE.thread_llm_judge ||
+    rule.type === EVALUATORS_RULE_TYPE.span_llm_judge
+  ) {
+    return rule.code.schema?.map((s) => s.name) || [];
+  }
+  if (
+    rule.type === EVALUATORS_RULE_TYPE.python_code ||
+    rule.type === EVALUATORS_RULE_TYPE.thread_python_code ||
+    rule.type === EVALUATORS_RULE_TYPE.span_python_code
+  ) {
+    return rule.code.metric ? [rule.code.metric] : [];
+  }
+  return [];
+};
+
+interface PlaygroundOutputScoresProps {
+  traceId: string | null;
+  selectedRuleIds: string[] | null | undefined;
```

Comment:
Perhaps selectedRuleIds should be a Set, rather than an array (Noticed that you used Set, which in favor of, as a big fan of performance & algorithms)

The .includes is expensive in a loop (Assuming there will be a lot of rules in the future)

If we can assume it's a small number, then this concern is a future trouble, and you can dismiss the suggestion.

As for no duplicates, prefer Set over array, that's better for maintenance / declared entity type.

Reply:
Seems better to stay with an array. Performance would be ok because it's not many rules